lib/internal/modules/esm/assets/load.js

```diff
@@ -0,0 +1,225 @@
+              BufferConcat(buffers, size)
+            );
+            f(body);
+          });
+        }),
+      };
+      getCache.set(parsed.href, entry);
+      fulfill(entry);
+    });
+  });
+  getCache.set(parsed.href, result);
+  return result;
+}
+
+/**
+ * Returns if an address has local status by if it is going to a local
+ * interface or is an address resolved by DNS to be a local interface
+ * @param {string} hostname url.hostname to test
+ * @returns {Promise<boolean>}
+ */
+async function isLocalAddress(hostname) {
+  try {
+    if (StringPrototypeStartsWith(hostname, '[')) {
+      hostname = StringPrototypeSlice(hostname, 1, -1);
+    }
+    const addr = await dnsLookup(hostname, { verbatim: true });
+    if (addr.family === 4) {
+      return addr.address === '127.0.0.1';
+    } else if (addr.family === 6) {
+      return normalizeIPv6(addr.address) ===
+        '0000:0000:0000:0000:0000:0000:0000:0001';
+    }
+  } catch {}
+  return false;
+}
+
+/**
+ * Used to expand interface addresses to a canonical form for comparisons
+ * E.G. comparing ::1, 0::1, etc. to 0000:0000:0000:0000:0000:0000:0000:0001
+ * Due to IPv6 having multiple ways to represent a single address we need to
+ * normalize
+ * @param {string} str
+ * @returns {string}
+ */
+function normalizeIPv6(str) {
+  const parts = StringPrototypeSplit(str, ':');
+  let expansionIndex = -1;
+  for (let i = 0; i < parts.length; i++) {
+    const part = parts[i];
+    if (part === '' && expansionIndex === -1 && parts.length < 8) {
+      expansionIndex = i;
+      continue;
+    }
+    parts[i] = StringPrototypePadStart(part, 4, '0');
+  }
+  if (parts.length < 8) {
+    ArrayPrototypeSplice(
+      parts,
+      expansionIndex,
+      1,
+      ...ArrayFrom({ length: 8 - parts.length + 1 }, () => '0000')
+    );
+  }
+  return ArrayPrototypeJoin(parts, ':');
+}
+
+/**
+ * @typedef CacheEntry
+ * @property resolvedHREF {Promise<string> | string}
+ * @property headers {Record<string, string>}
+ * @property body {Promise<Buffer> | Buffer}
+ */
+
+/**
+ * Only for GET requests, other requests would need new Map
+ * HTTP cache semantics keep diff caches
+ *
+ * Maps HREF to pending cache entry
+ * @type {Map<string, Promise<CacheEntry> | CacheEntry>}
+ */
+let getCache = new SafeMap();
+
+/**
+ * Fetches a location with a shared cache following redirects.
+ * Does not respect HTTP cache headers.
+ *
+ * This splits the header and body Promises so that things only needing
+ * headers don't need to wait on the body.
+ * @param {URL} parsed
+ * @returns {ReturnType<typeof fetchWithRedirects>}
+ */
+function fetch(parsed) {
+  const { href } = parsed;
+  const existing = getCache.get(href);
+  if (existing) {
+    return existing;
+  }
+  if (parsed.protocol === 'http:') {
+    return isLocalAddress(parsed.hostname).then((is) => {
+      if (is !== true) {
+        throw new Error(`Cannot load non-local address ${href} from network using http: protocol, use https: instead`);
```

Comment:
A proper `NODE_ERR_*` here?

Reply:
yup, those aren't quite done yet since I'll have to add a few and want to do it once the rest is ironed out rather than have error codes pop in and out while we refactor